flink-runtime/src/main/java/org/apache/flink/runtime/state/SharedStateRegistryImpl.java

```diff
@@ -174,6 +181,20 @@ public void registerAll(
         }
     }
 
+    @Override
+    public void registerAllAfterRestored(CompletedCheckpoint checkpoint, RestoreMode mode) {
+        registerAll(checkpoint.getOperatorStates().values(), checkpoint.getCheckpointID());
+        // In NO_CLAIM and LEGACY restore modes, shared state of the initial checkpoints must be
+        // preserved. This is achieved by advancing highestRetainCheckpointID here, and then
+        // checking entry.createdByCheckpointID against it on checkpoint subsumption.
+        // In CLAIM restore mode, the shared state of the initial checkpoints must be
+        // discarded as soon as it becomes unused - so highestRetainCheckpointID is not updated.
+        if (mode != RestoreMode.CLAIM) {
```

Comment:
Although the initial checkpoint is fully copied in `NO_CLAIM` mode, only the **next** checkpoint can be "claimed". 
`SharedStateRegistry ` and `CompletedCheckpointStore` still refer to the initial checkpoint until it is subsumed.
So we must prevent it's artifacts from being deleted.

Reply:
"initial checkpoint is fully copied" and owned by the current job, why it can not follow the normal subsumed way of other checkpoints?

So how the first full checkpoint is different from the second (next) checkpoint?

Reply:
> "initial checkpoint is fully copied" and owned by the current job, why it can not follow the normal subsumed way of other checkpoints?

No, initial checkpoint is NOT owned by the current job. When performing a **new** checkpoint after recovery, initial state is copied (if needed) by state backends and that copied state is included into the **new** checkpoint.